// @openauthjs/openauth/dist/esm/ui/form.js
// src/ui/form.tsx
import { jsxDEV } from "hono/jsx/jsx-dev-runtime";
function FormAlert(props) {
  return /* @__PURE__ */ jsxDEV("div", {
    "data-component": "form-alert",
    "data-color": props.color,
    children: [
      /* @__PURE__ */ jsxDEV("svg", {
        "data-slot": "icon-success",
        xmlns: "http://www.w3.org/2000/svg",
        fill: "none",
        viewBox: "0 0 24 24",
        "stroke-width": "1.5",
        stroke: "currentColor",
        children: /* @__PURE__ */ jsxDEV("path", {
          "stroke-linecap": "round",
          "stroke-linejoin": "round",
          d: "M9 12.75 11.25 15 15 9.75M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
        }, undefined, false, undefined, this)
      }, undefined, false, undefined, this),
      /* @__PURE__ */ jsxDEV("svg", {
        "data-slot": "icon-danger",
        xmlns: "http://www.w3.org/2000/svg",
        fill: "none",
        viewBox: "0 0 24 24",
        "stroke-width": "1.5",
        stroke: "currentColor",
        children: /* @__PURE__ */ jsxDEV("path", {
          "stroke-linecap": "round",
          "stroke-linejoin": "round",
          d: "M12 9v3.75m9-.75a9 9 0 1 1-18 0 9 9 0 0 1 18 0Zm-9 3.75h.008v.008H12v-.008Z"
        }, undefined, false, undefined, this)
      }, undefined, false, undefined, this),
      /* @__PURE__ */ jsxDEV("span", {
        "data-slot": "message",
        children: props.message
      }, undefined, false, undefined, this)
    ]
  }, undefined, true, undefined, this);
}
export {
  FormAlert
};
